### DISCUSSION TITLE
Use tooltip to show translation

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [avm99963/translateselectedtext#15] [Ideas]
Hi! Thanks for this extensions. I would find it more useful if I saw the translation in a tooltip without having to have a separate window/panel open. Could something like that be implemented?

**Comment by avm99963** [comment]
Hi Ivan!

You're welcome, thank you for your suggestion! I'm sorry for replying so late, I've had this project on hold for a long time due to having very little time to maintain it, but I'm now back and trying to improve the codebase, fix some issues and explore how to develop it further :)

Your suggestion is something many people have also been suggesting over the years, and it is something that in my opinion would be great to implement.

I've already thought of this, and some obstacles that I've encountered while trying to implement this feature are the following:

1. In order to do this, we could for instance inject the tooltip directly into the website, but that would require the `activeTab` permission in order to run a content script in the context of the tab.[1] The problem is that the cornerstone of the extension is to request the smallest subset of permissions possible, so although the `activeTab` permission is invisible to the end-user and would only allow the extension to access pages where translations are needed, we would have to assess whether this goes against the spirit of the extension (currently the extension only has access to the selected text, while doing this would give the extension access to the entire page).
2. Even if we managed to somehow display the translation (some years ago I experimented with showing it in a notification, although this has many drawbacks) we need to send the text to some API endpoint in order to get the translated text back. This introduces 2 problems I haven't been able to solve:
    - In order for the extension to contact the API endpoint, we would need to add a permission to access that origin. This clearly goes against the spirit of the extension.
    - When I looked into this some years ago, I didn't find any translation API endpoints free of charge (which is absolutely normal and expected). Another possibility to explore would be to run an open source translation service in a VPS, which I would have to pay for... But this is another obstacle.
        - When I developed the notifications prototype, I used a Google Translate endpoint similar to the one explained in [this blog post](https://www.labnol.org/code/19909-google-translate-api). The thing is using this endpoint in the extension goes against Google's terms of service.

So the main obstacle is the fact that I want the extension to request the least amount of permissions possible. The extension requests some permissions such as the "notifications" permission which are not essential to the core functionality of the extension, but none of them are privacy-invasive.

I would like to explore once again how to deal with these obstacles and implement this, so I'll give it a thought over the following weeks and see if I come up with something! If you have any ideas too or any other comment, please feel free to tell me :)

Again, thank you very much for sending the suggestion!

\[1]: [The activeTab permission - Chrome Developers](https://developer.chrome.com/docs/extensions/mv2/manifest/activeTab/)

**Comment by avm99963** [comment]
I still have to give this a thought, but for now I'll move this issue to the discussions tab so we can discuss about this there.
